src/libraries/Microsoft.Extensions.Caching.Abstractions/src/CacheEntryExtensions.cs

```diff
@@ -174,7 +174,8 @@ public static ICacheEntry SetOptions(this ICacheEntry entry, MemoryCacheEntryOpt
 
             foreach (PostEvictionCallbackRegistration postEvictionCallback in options.PostEvictionCallbacks)
             {
-                entry.RegisterPostEvictionCallback(postEvictionCallback.EvictionCallback!, postEvictionCallback.State);
+                ArgumentNullException.ThrowIfNull(postEvictionCallback.EvictionCallback);
```

Comment:
Why is this a beneficial change?  It seems like it's adding another check-and-throw that's just duplicating what RegisterPostEvictionCallback is about to do, anyway.

Reply:
I'm not opposite to simply remove `!`

Reply:
Here is a small discussion https://github.com/dotnet/runtime/issues/64147#issuecomment-1538380420

Reply:
`EvictionCallback` is typed as `PostEvictionDelegate?`.  `RegisterPostEvictionCallback` is typed to accept `PostEvictionDelegate`.  So we can't just remove the `!`.  I'm just not clear on why we need to change anything here.

Reply:
If it's fine too just pass null to the next check in RegisterPostEvictionCallback, I'll update it

Reply:
My original thought was to throw a more appropriate exception here if `postEvictionCallback.EvictionCallback` is null. `ArgumentNullException` isn't appropriate for this situation because there wasn't a `null` argument to this method.

Reply:
Here is the code that will get into this situation:

```C#
ICacheEntry entry = ...;

MemoryCacheEntryOptions options = new MemoryCacheEntryOptions();
options.PostEvictionCallbacks.Add(new PostEvictionCallbackRegistration());

entry.SetOptions(options);
```

Getting an `ArgumentNullException` from the above code doesn't make much sense IMO.

Reply:
Does something like `throw new ArgumentException("Eviction callbacks are not allowed to be null here, but one of the PostEvictionCallbacks was", nameof(options))` looks better?

Reply:
I would clean the message up a bit, but yes I think an `ArgumentException` on `options` is a better experience. Maybe something like:

"MemoryCacheEntryOptions.PostEvictionCallbacks contains a PostEvictionCallbackRegistration with a `null` EvictionCallback at index {0}."

We should also add a test.